### PR TITLE
2-metadata-importation-doc

### PR DESCRIPTION
**Pull request by ghost**
Add: Documentation about the metadata importation

### CODE REVIEW

**Review by rngadam — CHANGES_REQUESTED** [review]
We discussed in person to be more specific with actors.

Probably want to identify more clearly BEFORE processes and AFTER processes as the sequence diagrams you're presenting seems to be applicable to current situation.

instead of putting yaml embedded in the doc, create an example structure and a pydantic validator (with auto-generated doc)

**Comment by FrancoisWerbrouck-CFIA** [comment]
We discussed in person. 

The documentation needs to be reworked to offer a better understanding of the upcoming process within Nachet.

**Comment by rngadam** [comment]
should also add the github workflow checks as I see a couple basic issues that are detected by the checks:

https://github.com/ai-cfia/github-workflows

**Comment by FrancoisWerbrouck-CFIA** [comment]
I showed my doc to @JolanThomassin and took his feedback. I've then added small description to the graphs.

**Comment by JolanThomassin** [comment]
Very good graphs 👍

**Comment by rngadam** [comment]
> should also add the github workflow checks as I see a couple basic issues that are detected by the checks:
> 
> https://github.com/ai-cfia/github-workflows

@FrancoisWerbrouck-CFIA I see a thumbs up but what we need is a reference to a new issue describing the work to be done to ensure the work is properly scheduled.

File: file-validator.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    userData: UserData 
+    
+class ClientFeedback(BaseModel):
+    correctIdentification: bool
+    historicalComparison: str
+    
+    
+def is_yaml_file(file_path):
+    _, file_extension = os.path.splitext(file_path)
+    return file_extension.lower() == '.yaml'
```
Comment (rngadam) [inline file-validator.py:62]: yml is also a very common extension

more reliable to look at headers than extension to determine file types:

https://github.com/ahupp/python-magic

File: file-validator.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+def buildIndex(output:str):
+    # Ask for the url in the console
+
+    # # Retrieve some client info from the URL
+
+    # Ask Data (I dont have access to the data url atm)
+    clientEmail=input("clientEmail: ")
+    clientExpertise= input("Expertise: ")
+    clientData=ClientData(clientEmail=clientEmail,clientExpertise=clientExpertise)
```
Comment (rngadam) [inline file-validator.py:84]: you're inconsistent in how you write your code. here the whitespaces before adn after the = signs (the rule is spaces except for parameters to fucntions). ruff extension should be able to give you feedback here.

File: file-validator.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    
+    
+def openIndex(path):
+    if is_yaml_file(path):
+        # Load YAML template
+        with open(path, 'r') as file:
+            yaml_data = yaml.safe_load(file)
+    try: # Validate and parse YAML data using Pydantic model
+        data = Index(**yaml_data)
+    except Exception as error: # Errtor raised by Pydantic Validator
```
Comment (rngadam) [inline file-validator.py:145]: Catch the specific exception, not every type of exception. do not want to hide other type of exceptions.

File: file-validator.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        print(error)
```
Comment (rngadam) [inline file-validator.py:146]: should the program really continues without failing? typically should re-raise your custom exception to ensure the caller handles failure here.

**Review by rngadam — CHANGES_REQUESTED** [review]

File: nachetDb-1.0.0/db-creation.py [hunk 5]
```diff
@@ -18,34 +18,34 @@
...
-#     )
-# """ % ("nachetdb_1.0.0"))
+# Create Pictures table
+cur.execute("""
+    CREATE TABLE  \"%s\".pictures (
+        id uuid  PRIMARY KEY,
+        picture JSON,
+        indexID uuid REFERENCES "nachetdb_1.0.0".indexes(id)
+    )
+""" % ("nachetdb_1.0.0"))
```
Comment (rngadam) [inline nachetDb-1.0.0/db-creation.py:37]: don't get into the habit of bypassing the built in parameters facilities

https://www.psycopg.org/psycopg3/docs/basic/params.html

If the variables containing the data to send to the database come from an untrusted source (such as data coming from a form on a web site) an attacker could easily craft a malformed string, either gaining access to unauthorized data or performing destructive operations on the database. This form of attack is called [SQL injection](https://en.wikipedia.org/wiki/SQL_injection) and is known to be one of the most widespread forms of attack on database systems. Before continuing, please print [this page](https://xkcd.com/327/) as a memo and hang it onto your desk.

File: nachetDb-1.0.0/db-creation.py [hunk 6]
```diff
@@ -18,34 +18,34 @@
...
-#     )
-# """ % ("nachetdb_1.0.0"))
+# Create Indexes table
+cur.execute("""
+    CREATE TABLE  \"%s\".indexes (
+        id uuid  PRIMARY KEY,
+        index JSON,
+        ownerID uuid REFERENCES "nachetdb_1.0.0".users(id)
+    )
+""" % ("nachetdb_1.0.0"))
```
Comment (rngadam) [inline nachetDb-1.0.0/db-creation.py:28]: keep 1.0.0 for the first production version. start with 0.0.1

File: nachetDb-1.0.0/deployment-mass-import.py [hunk 7]
```diff
@@ -462,14 +462,17 @@ def uploadIndexDB(path:str,userID:str):
     with open(path, 'r') as file:
         data = file.read()
 
+    #Build indexID
+    indexID = uuid.uuid4()
+
     # Execute the INSERT statement
     print(userID)
-    cur.execute("INSERT INTO \"nachetdb_1.0.0\".indexes (index, ownerID) VALUES (%s, %s)", (data, userID))
+    cur.execute("INSERT INTO \"nachetdb_1.0.0\".indexes (id,index, ownerID) VALUES (%s,%s, %s)", (indexID,data, userID))
```
Comment (rngadam) [inline nachetDb-1.0.0/deployment-mass-import.py:470]: your spacing throughout is inconsistent

**Review by rngadam — CHANGES_REQUESTED** [review]
Why have two different PR open?

File: nachetDb-1.0.0/db-creation.py (hunk 5 above)
Comment (Francois-Werbrouck) [reply nachetDb-1.0.0/db-creation.py:37]: You cant use parameters query for schema, table or column name. I understand the need to protect from injection and will apply the parameters query for my future cursor.

Comment (rngadam) [reply nachetDb-1.0.0/db-creation.py:37]: As discussed yesterday, use ```set search_path``` for DDL operations

https://www.postgresql.org/docs/current/ddl-schemas.html

>Qualified names are tedious to write, and it's often best not to wire a particular schema name into applications anyway. Therefore tables are often referred to by unqualified names, which consist of just the table name. The system determines which table is meant by following a search path, which is a list of schemas to look in. The first matching table in the search path is taken to be the one wanted. If there is no match in the search path, an error is reported, even if matching table names exist in other schemas in the database.

File: metadata-doc.md [hunk 8]
```diff
@@ -278,14 +342,15 @@ erDiagram
     string container  
   }
   indexes{
```
Comment (rngadam) [inline metadata-doc.md:358]: index.yaml (here indexes) is a client-side concept to facilitate reducing the number of properties we have to input to import an image, not really something we'd expect to be in the API or database.

**Review by rngadam — CHANGES_REQUESTED** [review]

**Comment by rngadam** [comment]
@Francois-Werbrouck my previous opened conversations are not acknowledged or commented?

File: metadata-doc.md (hunk 8 above)
Comment (Francois-Werbrouck) [reply metadata-doc.md:358]: Fixed in #11 -> https://github.com/ai-cfia/nachet-metadata-format/commit/81ff9b99a16d744d8afb4ee631128213ddb586da

**Comment by Francois-Werbrouck** [comment]
> @Francois-Werbrouck my previous opened conversations are not acknowledged or commented?

@rngadam The changes have already been made in other versions and then improved upon. (Index -> Session -> picture_set)

Having old PRs waiting for review with already alot of increment and changes into the structure I've applied the latest changes to the latest version #13

**Review by rngadam — APPROVED** [review]
I'll follow up in your newer PR

**Pull request closed without merging**